crates/blockchain/blockchain.rs

```diff
@@ -161,6 +143,120 @@ pub fn validate_block(
     Ok(())
 }
 
+pub fn is_canonical(
+    store: &Store,
+    block_number: BlockNumber,
+    block_hash: BlockHash,
+) -> Result<bool, StoreError> {
+    match store.get_canonical_block_hash(block_number)? {
+        Some(hash) if hash == block_hash => Ok(true),
+        _ => Ok(false),
+    }
+}
+
+/// Applies new fork choice data to the current blockchain. It performs validity checks:
+/// - The finalized, safe and head hashes must correspond to already saved blocks.
+/// - The saved blocks should be in the correct order (finalized <= safe <= head).
+/// - They must be connected.
+///
+/// After the validity checks, the canonical chain is updated so that all head's ancestors
+/// and itself are made canonical.
+pub fn apply_fork_choice(
+    store: &Store,
+    head_hash: H256,
+    safe_hash: H256,
+    finalized_hash: H256,
+) -> Result<(), InvalidForkChoice> {
+    if head_hash.is_zero() {
+        return Err(InvalidForkChoice::InvalidHeadHash);
+    }
+
+    // We get the block bodies even if we only use headers them so we check that they are
+    // stored too.
+    let finalized_header_res = store.get_block_by_hash(finalized_hash).map_err(wrap)?;
```

Comment:
nit: do `store.get_block_by_hash(finalized_hash)` and 

implement the trait `Into` from `StoreError` to `ChainError`

Reply:
Turns out this wasn't really necessary. We just the "From" trait is automatically implemented when deriving from `Error`.